Second Hands On
Humaira Akhtari
New York City House Prices Regression Model

In [24]:
#Importing necessary libraries
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.compose import ColumnTransformer
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn import tree
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.linear_model import Ridge, Lasso, ElasticNet
from sklearn.model_selection import cross_val_score, KFold
from sklearn.metrics import make_scorer, mean_squared_error
import numpy as np

In [6]:
hprice_reg=pd.read_csv('/content/NY-House-Dataset.csv')
hprice_reg.head()

BROKERTITLE                TYPE  \
0        Brokered by Douglas Elliman  -111 Fifth Ave      Condo for sale   
1                                Brokered by Serhant      Condo for sale   
2                             Brokered by Sowae Corp      House for sale   
3                                Brokered by COMPASS      Condo for sale   
4  Brokered by Sotheby's International Realty - E...  Townhouse for sale   

       PRICE  BEDS       BATH  PROPERTYSQFT  \
0     315000     2   2.000000        1400.0   
1  195000000     7  10.000000       17545.0   
2     260000     4   2.000000        2015.0   
3      69000     3   1.000000         445.0   
4   55000000     7   2.373861       14175.0   

                                             ADDRESS                    STATE  \
0                               2 E 55th St Unit 803       New York, NY 10022   
1  Central Park Tower Penthouse-217 W 57th New Yo...       New York, NY 10019   
2                                   620 Sinclair Ave  Staten Island, NY 10312   
3                            2 E 55th St Unit 908W33      Manhattan, NY 10022   
4                                        5 E 64th St       New York, NY 10065   

                                        MAIN_ADDRESS  \
0             2 E 55th St Unit 803New York, NY 10022   
1  Central Park Tower Penthouse-217 W 57th New Yo...   
2            620 Sinclair AveStaten Island, NY 10312   
3         2 E 55th St Unit 908W33Manhattan, NY 10022   
4                      5 E 64th StNew York, NY 10065   

  ADMINISTRATIVE_AREA_LEVEL_2  LOCALITY      SUBLOCALITY       STREET_NAME  \
0             New York County  New York        Manhattan  East 55th Street   
1               United States  New York  New York County          New York   
2               United States  New York  Richmond County     Staten Island   
3               United States  New York  New York County          New York   
4               United States  New York  New York County          New York   

          LONG_NAME                                  FORMATTED_ADDRESS  \
0   Regis Residence  Regis Residence, 2 E 55th St #803, New York, N...   
1  West 57th Street             217 W 57th St, New York, NY 10019, USA   
2   Sinclair Avenue     620 Sinclair Ave, Staten Island, NY 10312, USA   
3  East 55th Street               2 E 55th St, New York, NY 10022, USA   
4  East 64th Street               5 E 64th St, New York, NY 10065, USA   

    LATITUDE  LONGITUDE  
0  40.761255 -73.974483  
1  40.766393 -73.980991  
2  40.541805 -74.196109  
3  40.761398 -73.974613  
4  40.767224 -73.969856

In [7]:
hprice_reg.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4801 entries, 0 to 4800
Data columns (total 17 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   BROKERTITLE                  4801 non-null   object 
 1   TYPE                         4801 non-null   object 
 2   PRICE                        4801 non-null   int64  
 3   BEDS                         4801 non-null   int64  
 4   BATH                         4801 non-null   float64
 5   PROPERTYSQFT                 4801 non-null   float64
 6   ADDRESS                      4801 non-null   object 
 7   STATE                        4801 non-null   object 
 8   MAIN_ADDRESS                 4801 non-null   object 
 9   ADMINISTRATIVE_AREA_LEVEL_2  4801 non-null   object 
 10  LOCALITY                     4801 non-null   object 
 11  SUBLOCALITY                  4801 non-null   object 
 12  STREET_NAME                  4801 non-null   object 
 13  LONG_NAME         

In [8]:
hprice_reg.dropna(inplace=True)
target = 'PRICE'

In [15]:
# define X and y
X = hprice_reg.drop(target, axis=1)
y = hprice_reg[target]
# Selecting only numerical features
numerical_features = X.select_dtypes(include=np.number).columns
X_numerical = X[numerical_features]
# Scale the features for better performance of regression models
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X_numerical)
#Create new data frame for scaled numerical features
X_scaled = pd.DataFrame(X_scaled, columns=numerical_features, index=X.index)

# Concatenate scaled numerical features with non-numerical features
X_processed = pd.concat([X.drop(columns=numerical_features), X_scaled], axis=1)


In [17]:
##Making the split using 30 70 ratio and setting the random state to 42
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.3, random_state=42)


These numbers are too high I do not know what I am doing wrong.

In [18]:
# Initialize and train the Linear Regression model
linear_model = LinearRegression()
linear_model.fit(X_train, y_train)

# Predict on the testing set
y_pred_linear = linear_model.predict(X_test)

# Evaluate the Linear Regression model
print("Linear Regression Mean Squared Error:", mean_squared_error(y_test, y_pred_linear))
print("Linear Regression R^2 Score:", r2_score(y_test, y_pred_linear))


Linear Regression Mean Squared Error: 37137447259016.51
Linear Regression R^2 Score: -0.8794377800042665


In [20]:
# Initialize and train the Random Forest Regressor
rf_model = RandomForestRegressor(n_estimators=300, random_state=42)
rf_model.fit(X_train, y_train)

# Predict on the testing set
y_pred_rf = rf_model.predict(X_test)

# Evaluate the Random Forest Regressor
print("Random Forest Regressor Mean Squared Error:", mean_squared_error(y_test, y_pred_rf))
print("Random Forest Regressor R^2 Score:", r2_score(y_test, y_pred_rf))


Random Forest Regressor Mean Squared Error: 15768310537878.074
Random Forest Regressor R^2 Score: 0.20200334866763614


In [21]:
# Create a RandomForestRegressor with regularization parameters
regressor = RandomForestRegressor(
    n_estimators=100,          # Number of trees
    max_depth=10,              # Limit the depth of each tree
    min_samples_split=5,       # Minimum samples required to split an internal node
    min_samples_leaf=4,        # Minimum samples required to be at a leaf node
    max_features='sqrt',       # Use the square root of the total features at each split
    random_state=42
)

# Fit the model
regressor.fit(X_train, y_train)

# Make predictions and evaluate
y_pred = regressor.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
print("Random Forest Regressor Mean Squared Error:", mean_squared_error(y_test, y_pred))
print("Random Forest Regressor R^2 Score:", r2_score(y_test, y_pred_rf))


Random Forest Regressor Mean Squared Error: 50920298144478.24
Random Forest Regressor R^2 Score: 0.20200334866763614


In [31]:
categorical_cols = ['TYPE', 'ADDRESS', 'STATE', 'MAIN_ADDRESS', 'ADMINISTRATIVE_AREA_LEVEL_2', 'STREET_NAME']
numerical_cols = ['PRICE', 'BEDS', 'BATH', 'PROPERTYSQFT', 'LATITUDE', 'LONGITUDE']
# One-hot encode categorical features
encoder = OneHotEncoder(drop='first', sparse_output=False)
encoded_data = encoder.fit_transform(hprice_reg[categorical_cols])

# Create a DataFrame from encoded data
encoded_columns = encoder.get_feature_names_out(categorical_cols)
df_encoded = pd.DataFrame(encoded_data, columns=encoded_columns, index=hprice_reg.index)

# Combine numerical data and encoded categorical data
X = pd.concat([hprice_reg[numerical_cols], df_encoded], axis=1)
y = hprice_reg['PRICE']
df_encoded

TYPE_Coming Soon  TYPE_Condo for sale  TYPE_Condop for sale  \
0                  0.0                  1.0                   0.0   
1                  0.0                  1.0                   0.0   
2                  0.0                  0.0                   0.0   
3                  0.0                  1.0                   0.0   
4                  0.0                  0.0                   0.0   
...                ...                  ...                   ...   
4796               0.0                  0.0                   0.0   
4797               0.0                  0.0                   0.0   
4798               0.0                  0.0                   0.0   
4799               0.0                  1.0                   0.0   
4800               0.0                  0.0                   0.0   

      TYPE_Contingent  TYPE_For sale  TYPE_Foreclosure  TYPE_House for sale  \
0                 0.0            0.0               0.0                  0.0   
1                 0.0            0.0               0.0                  0.0   
2                 0.0            0.0               0.0                  1.0   
3                 0.0            0.0               0.0                  0.0   
4                 0.0            0.0               0.0                  0.0   
...               ...            ...               ...                  ...   
4796              0.0            0.0               0.0                  0.0   
4797              0.0            0.0               0.0                  0.0   
4798              0.0            0.0               0.0                  0.0   
4799              0.0            0.0               0.0                  0.0   
4800              0.0            0.0               0.0                  0.0   

      TYPE_Land for sale  TYPE_Mobile house for sale  \
0                    0.0                         0.0   
1                    0.0                         0.0   
2                    0.0                         0.0   
3                    0.0                         0.0   
4                    0.0                         0.0   
...                  ...                         ...   
4796                 0.0                         0.0   
4797                 0.0                         0.0   
4798                 0.0                         0.0   
4799                 0.0                         0.0   
4800                 0.0                         0.0   

      TYPE_Multi-family home for sale  ...  STREET_NAME_West 65th Street  \
0                                 0.0  ...                           0.0   
1                                 0.0  ...                           0.0   
2                                 0.0  ...                           0.0   
3                                 0.0  ...                           0.0   
4                                 0.0  ...                           0.0   
...                               ...  ...                           ...   
4796                              0.0  ...                           0.0   
4797                              0.0  ...                           0.0   
4798                              0.0  ...                           0.0   
4799                              0.0  ...                           0.0   
4800                              0.0  ...                           0.0   

      STREET_NAME_West Brighton  STREET_NAME_West Bronx  \
0                           0.0                     0.0   
1                           0.0                     0.0   
2                           0.0                     0.0   
3                           0.0                     0.0   
4                           0.0                     0.0   
...                         ...                     ...   
4796                        0.0                     0.0   
4797                        0.0                     0.0   
4798                        0.0                     0.0   
4799                        0.0                     

Still Getting very high numbers. Can not figure out the issue.

In [34]:
# Assuming X and y are the features
# Define the model with regularization parameters
regressor = RandomForestRegressor(
    n_estimators=300,
    max_depth=5,
    min_samples_split=5,
    min_samples_leaf=4,
    max_features='sqrt',
    random_state=42
)

# Set up k-fold cross-validation (5 folds)
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Perform cross-validation and calculate MSE for each fold
scores = cross_val_score(
    regressor, X, y, cv=kf, scoring=make_scorer(mean_squared_error)
)

# Calculate the mean and standard deviation of the MSE scores
mean_mse = np.mean(scores)
std_mse = np.std(scores)

print(f"Mean MSE from cross-validation: {mean_mse:.2f}")
print(f"R^2 Score: {r2_score(y_test, y_pred):.2f}")


Mean MSE from cross-validation: 976793928714606.25
R^2 Score: -1.58


Gradient Boosting Regressor is a machine learning algorithm that builds a strong model by sequentially adding small decision trees, each correcting the errors of the previous ones. It combines these trees to make accurate predictions for continuous values.



In [35]:
#Gradient Booster Regressor algorithm
#define the gradient boosting model with different parameters
gb_model = GradientBoostingRegressor(n_estimators=300, learning_rate=0.1, max_depth=3, random_state=42)
gb_model.fit(X_train, y_train)

# Predict on the testing set
y_pred_gb = gb_model.predict(X_test)

# Evaluate the Gradient Boosting Regressor
print("Gradient Boosting Regressor Mean Squared Error:", mean_squared_error(y_test, y_pred_gb))
print("Gradient Boosting Regressor R^2 Score:", r2_score(y_test, y_pred_gb))


Gradient Boosting Regressor Mean Squared Error: 20688458402353.9
Gradient Boosting Regressor R^2 Score: -0.04699361968101878


OTHER REGRESSION MODELS:

Ridge Regression (L2 Regularization)
Lasso Regression (L1 Regularization)
Elastic Net (Combination of L1 and L2)

In [37]:
# L2 Regularization: Ridge Regression
ridge = Ridge(alpha=1.0)  # alpha controls the regularization strength; higher means more regularization
ridge.fit(X_train, y_train)
ridge_predictions = ridge.predict(X_test)
print("Ridge MSE:", mean_squared_error(y_test, ridge_predictions))
print("Ridge R^2 Score:", r2_score(y_test, ridge_predictions))

# L1 Regularization: Lasso Regression
lasso = Lasso(alpha=0.1)  # alpha is the regularization parameter
lasso.fit(X_train, y_train)
lasso_predictions = lasso.predict(X_test)
print("Lasso MSE:", mean_squared_error(y_test, lasso_predictions))
print("Lasso R^2 Score:", r2_score(y_test, lasso_predictions))

# L1 + L2 Regularization: Elastic Net
# l1_ratio balances between L1 and L2 (0 = pure L2, 1 = pure L1)
elastic_net = ElasticNet(alpha=0.1, l1_ratio=0.5)
elastic_net.fit(X_train, y_train)
elastic_net_predictions = elastic_net.predict(X_test)
print("Elastic Net MSE:", mean_squared_error(y_test, elastic_net_predictions))
print("Elastic Net R^2 Score:", r2_score(y_test, elastic_net_predictions))

Ridge MSE: 37113544073466.336
Ridge R^2 Score: -0.8782280966980307
Lasso MSE: 37137445326277.89
Lasso R^2 Score: -0.8794376821929668
Elastic Net MSE: 33631139959068.508
Elastic Net R^2 Score: -0.7019919162143937


CONCLUSION:

Linear Regression - 0.879
Random Forest Regressor - 0.20

Random Forest Regressor with regularization - 0.20

Gradient Boosting Regressor - 0.046

Ridge Regression - 0.87

Lasso Regression - 0.87

Elastic Regression - 0.70

Based on these R² scores, we can evaluate and rank the performance of each model.

1. Linear Regression: 0.879
2. Ridge Regression: 0.87
3. Lasso Regression: 0.87
4. Elastic Net Regression: 0.70
5. Random Forest Regressor: 0.20 (both with and without regularization)
6. Gradient Boosting Regressor: 0.046
